README
NAME
    Games::PuzzleCars - Puzzle game involving cars

SYNOPSIS
        perl puzzlecars.pl

    Options:
        --easy    Use the easy map
        --hard    Use the hard map

INSTRUCTIONS
    Click on the intersections to change which way the cars will turn.  Try to
    match the cars' color with the road they use to exit the screen to earn
    points.  If the colors don't match, you lose points.

PREREQUISIES
    * SDL
    * Mouse


Changes
Revision history for Games-PuzzleCars

{{$NEXT}}
    - Initial release



lib/Games/PuzzleCars/Car.pm
package Games::PuzzleCars::Car;
use Mouse;
use Math::Trig qw( deg2rad );
use Games::PuzzleCars::Sprite;

has [qw( x y rot v_x v_y )] => (
    is       => 'rw',
    isa      => 'Num',
    required => 1,
);

has direction => (
    is       => 'rw',
    isa      => 'Str',
    required => 1,
);

has turn => (
    is      => 'rw',
    isa     => 'HashRef',
    clearer => '_clear_turn',
);

has turned => (
    is      => 'rw',
    isa     => 'Bool',
    default => 0,
);

has map => (
    is       => 'ro',
    isa      => 'Games::PuzzleCars::Map',
    required => 1,
);

has road => (
    is       => 'rw',
    isa      => 'Games::PuzzleCars::Road',
    required => 1,
);

has next_road => (
    is      => 'rw',
    isa     => 'Games::PuzzleCars::Road',
    clearer => '_clear_next_road',
);

has _sprite => (
    is       => 'ro',
    isa      => 'Games::PuzzleCars::Sprite',
    required => 1,
    handles  => [qw( draw )],
);

has color => (
    is       => 'ro',
    isa      => 'Str',
    required => 1,
);

around BUILDARGS => sub {
    my ( $orig, $class, %args ) = @_;

    my $sprite = Games::PuzzleCars::Sprite->new(
        rect    => $args{rect},
        surface => $args{surface},
    );

    return $class->$orig( _sprite => $sprite, %args );
};

before draw => sub {
    my ($self) = @_;

    my $sprite = $self->_sprite;
    my $rect   = $sprite->rect;

    $sprite->x( $self->x - $rect->w / 2 );
    $sprite->y( $self->y - $rect->h / 2 );
    $sprite->rotation( $self->rot );
};

sub move {
    my ( $self, $step, $c, $t ) = @_;

    if ( !$self->turned && ( my $turn = $self->turn ) ) {
        $turn->{angle} += $turn->{delta} * $step;

        my ( $xc, $yc, $r, $angle ) = @$turn{qw( x y r angle )};

        my $delta_dir = $turn->{delta} <=> 0;

        my $x = $xc + cos( deg2rad($angle) ) * $r;
        my $y = $yc - sin( deg2rad($angle) ) * $r;

        $self->x($x);
        $self->y($y);
        $self->rot( $angle + 90 * $delta_dir );

        if ( ( $turn->{angle} <=> $turn->{max_angle} ) == $delta_dir ) {
            my $finish = $turn->{finish};
            $self->x( $finish->{x} );
            $self->y( $finish->{y} );
            $self->rot( $turn->{max_angle} + 90 * $delta_dir );
            $self->v_x( $finish->{v_x} );
            $self->v_y( $finish->{v_y} );
            $self->direction( $finish->{direction} );
            $self->_clear_turn();
            $self->turned(1);
        }
    }
    else {
        $self->x( $self->x + $self->v_x * $step );
        $self->y( $self->y + $self->v_y * $step );
        $self->road->turn($self) unless $self->turned;
    }

    if (   $self->next_road
        && $self->next_road->contains( $self->x, $self->y ) )
    {
        my $road = $self->next_road;
        my $next = $road->next( $self->direction );

        $self->road($road);

        if ($next) {
            $self->next_road($next);
        }
        else {
            $self->_clear_next_road();
        }
        $self->turned(0);
    }
}

no Mouse;

1;


lib/Games/PuzzleCars/Map.pm
package Games::PuzzleCars::Map;
use Mouse;
use List::Util qw( max );
use SDL::Rect;
use SDL::Event;
use SDL::Events;
use SDLx::Surface;
use SDLx::Sprite;
use Games::PuzzleCars::Road;
use Games::PuzzleCars::Intersection;

has _data => (
    is       => 'ro',
    isa      => 'HashRef',
    required => 1,
);

has background => (
    is      => 'ro',
    isa     => 'SDLx::Surface',
    lazy    => 1,
    builder => '_build_background',
);

has roads => (
    is      => 'ro',
    isa     => 'ArrayRef[ArrayRef]',
    lazy    => 1,
    builder => '_build_roads',
);

has intersections => (
    is      => 'ro',
    isa     => 'ArrayRef[Games::PuzzleCars::Intersection]',
    lazy    => 1,
    builder => '_build_intersections',
);

has borders => (
    is      => 'ro',
    isa     => 'ArrayRef[Games::PuzzleCars::Road]',
    lazy    => 1,
    builder => '_build_borders',
);

has exits => (
    is      => 'ro',
    isa     => 'ArrayRef[ArrayRef]',
    lazy    => 1,
    builder => '_build_exits',
);

has exit_colors => (
    is      => 'ro',
    isa     => 'HashRef[Int]',
    default => sub { { r => 0x920A0AFF, g => 0x479932FF, y => 0xB9C043FF, } },
);

has [qw( w h tile_w tile_h )] => (
    is       => 'ro',
    isa      => 'Int',
    required => 1,
);

sub _build_background {
    my ($self) = @_;

    my $bg = SDLx::Surface->new(
        w     => $self->w * $self->tile_w,
        h     => $self->h * $self->tile_h,
        color => 0x000000FF,
    );

    my $data = $self->_data;

    my $road_sprite = SDLx::Sprite->new( image => $data->{roads}{image} );

    my $road_w       = $data->{roads}{w};
    my $road_h       = $data->{roads}{h};
    my $road_mapping = $data->{roads}{mapping};

    my $grid  = $data->{grid};
    my $exits = $data->{exits};

    my %horizontal = ( 0 => 'WEST',  1 => 'EAST' );
    my %vertical   = ( 0 => 'NORTH', 1 => 'SOUTH' );

    foreach my $col_id ( 0 .. $self->w - 1 ) {
        foreach my $row_id ( 0 .. $self->h - 1 ) {
            while ( my ( $y_i, $y_dir ) = each(%vertical) ) {
                while ( my ( $x_i, $x_dir ) = each(%horizontal) ) {

                    my $index = 0;
                    if ( my %directions = %{ $grid->[$col_id][$row_id] } ) {
                        $index += 1 if defined $directions{$y_dir};
                        $index += 2 if defined $directions{$x_dir};
                    }
                    else {
                        $index = 4;
                    }

                    my @tiles = @{ $road_mapping->[ 2 * $y_i + $x_i ] };

                    my @tile_group = @{ $tiles[$index] };
                    my @tile       = @{ $tile_group[ int rand @tile_group ] };
                    my $x          = $tile[0];
                    my $y          = $tile[1];

                    $road_sprite->clip(
                        [ $x * $road_w, $y * $road_h, $road_w, $road_h ] );
                    $road_sprite->draw_xy(
                        $bg,
                        ( 2 * $col_id + $x_i ) * $road_w,
                        ( 2 * $row_id + $y_i ) * $road_h
                    );
                }
            }
            if ( my $color = $exits->[$col_id][$row_id] ) {
                $bg->draw_circle_filled(
                    [   ( $col_id + 0.5 ) * $self->tile_w,
                        ( $row_id + 0.5 ) * $self->tile_h
                    ],
                    $road_h / 4,
                    $self->exit_colors->{$color}
                );
            }
        }
    }

    return $bg;
}

sub _build_roads {
    my ($self) = @_;

    my @roads;

    my $data = $self->_data;

    my $grid = $data->{grid};

    my $arrow = SDLx::Sprite->new( image => $data->{intersection}{image} );
    $arrow->alpha_key(0x000000);

    foreach my $col_id ( 0 .. $self->w - 1 ) {
        foreach my $row_id ( 0 .. $self->h - 1 ) {
            if ( my %directions = %{ $grid->[$col_id][$row_id] } ) {
                if ( keys %directions > 2 ) {
                    $roads[$col_id][$row_id]
                        = Games::PuzzleCars::Intersection->new(
                        map        => $self,
                        x          => $col_id,
                        y          => $row_id,
                        arrow      => $arrow,
                        directions => \%directions,
                        );
                }
                else {
                    $roads[$col_id][$row_id] = Games::PuzzleCars::Road->new(
                        map        => $self,
                        x          => $col_id,
                        y          => $row_id,
                        directions => \%directions,
                    );
                }
            }
        }
    }

    return \@roads;
}

sub _build_intersections {
    my ($self) = @_;

    return [
        grep { $_ && $_->isa('Games::PuzzleCars::Intersection') }
        map {@$_} @{ $self->roads }
    ];
}

sub _build_borders {
    my ($self) = @_;

    return [
        grep {
            $_
                && ( $_->x == 0
                || $_->x == $self->w - 1
                || $_->y == 0
                || $_->y == $self->h - 1 )
            } map {@$_} @{ $self->roads }
    ];
}

sub _build_exits {
    my ($self) = @_;
    return $self->_data->{exits};
}

around BUILDARGS => sub {
    my ( $orig, $class, %args ) = @_;

    open my $map_file, '<', $args{file};
    my @map = map { chomp; [ split //, $_ ] } <$map_file>;

    my @grid;
    my @exits;

    my $h = $#map;
    my $w = max map { $#{$_} } @map;

    foreach my $row_id ( 0 .. $h ) {
        foreach my $col_id ( 0 .. $w ) {

            my ( $prev_row, $next_row ) = ( $row_id - 1, $row_id + 1 );
            $prev_row = $row_id + 1 if $row_id == 0;
            $next_row = $row_id - 1 if $row_id == $h;
            my @v = ( $map[$prev_row][$col_id], $map[$next_row][$col_id] );

            my ( $prev_col, $next_col ) = ( $col_id - 1, $col_id + 1 );
            $prev_col = $col_id + 1 if $col_id == 0;
            $next_col = $col_id - 1 if $col_id == $w;
            my @h = ( $map[$row_id][$prev_col], $map[$row_id][$next_col] );

            my $cell = $map[$row_id][$col_id];

            if ( $cell =~ /[Rrgy]/ ) {
                my %directions;
                $directions{WEST}  = 1 if $h[0] =~ /[Rrgy]/;
                $directions{EAST}  = 1 if $h[1] =~ /[Rrgy]/;
                $directions{NORTH} = 1 if $v[0] =~ /[Rrgy]/;
                $directions{SOUTH} = 1 if $v[1] =~ /[Rrgy]/;

                $grid[$col_id][$row_id] = \%directions;
                $exits[$col_id][$row_id] = $cell if $cell =~ /[rgy]/;
            }
            else {
                $grid[$col_id][$row_id] = {};
            }
        }
    }

    return $class->$orig(
        _data => {
            grid         => \@grid,
            exits        => \@exits,
            roads        => $args{roads},
            intersection => $args{intersection},
        },
        w      => $w + 1,
        h      => $h + 1,
        tile_w => $args{roads}{w} * 2,
        tile_h => $args{roads}{h} * 2,
    );
};

sub handle_event {
    my ( $self, $event ) = @_;

    return unless $event->type == SDL_MOUSEBUTTONDOWN;

    $_->handle_event($event) foreach @{ $self->intersections };
}

sub draw {
    my ( $self, $surface ) = @_;

    $self->background->blit($surface);
    $_->draw($surface) foreach @{ $self->intersections };
}

no Mouse;

1;


lib/Games/PuzzleCars/Road.pm
package Games::PuzzleCars::Road;
use Mouse;

has map => (
    is       => 'ro',
    isa      => 'Games::PuzzleCars::Map',
    required => 1,
);

has [qw( x y )] => (
    is       => 'ro',
    isa      => 'Int',
    required => 1,
);

has directions => (
    is       => 'ro',
    isa      => 'HashRef',
    required => 1,
);

sub _next_direction {
    my ( $self, $direction ) = @_;

    if ( !defined $self->directions->{$direction} ) {
        my %opposite = (
            WEST  => 'EAST',
            EAST  => 'WEST',
            NORTH => 'SOUTH',
            SOUTH => 'NORTH',
        );

        ($direction) = grep { $_ ne $opposite{$direction} }
            keys %{ $self->directions };
    }

    return $direction;
}

sub next {
    my ( $self, $direction ) = @_;

    $direction = $self->_next_direction($direction);

    my $x = $self->x;
    my $y = $self->y;
    $x -= 1 if $direction eq 'WEST';
    $x += 1 if $direction eq 'EAST';
    $y -= 1 if $direction eq 'NORTH';
    $y += 1 if $direction eq 'SOUTH';

    my $roads = $self->map->roads;
    return undef unless defined $roads->[$x][$y];

    return $roads->[$x][$y];
}

sub turn {
    my ( $self, $car ) = @_;

    my $d  = $car->direction;
    my $nd = $self->_next_direction($d);

    return if $d eq $nd;

    my $x = $car->x;
    my $y = $car->y;

    my $left   = $self->x * $self->map->tile_w + 25;
    my $right  = $self->x * $self->map->tile_w + 75;
    my $top    = $self->y * $self->map->tile_h + 25;
    my $bottom = $self->y * $self->map->tile_h + 75;

    my $r_i     = 12;
    my $r_o     = 38;
    my $delta_i = 4;
    my $delta_o = 2;

    my ( $xc, $yc, $r, $delta, $angle, $max_angle );
    if ( $d eq 'WEST' && $x < $right ) {
        $xc = $right;
        if ( $nd eq 'NORTH' ) {
            $yc        = $top;
            $r         = $r_i;
            $delta     = -$delta_i;
            $angle     = 270;
            $max_angle = 180;
        }
        else {
            $yc        = $bottom;
            $r         = $r_o;
            $delta     = $delta_o;
            $angle     = 90;
            $max_angle = 180;
        }
    }
    elsif ( $d eq 'EAST' && $x > $left ) {
        $xc = $left;
        if ( $nd eq 'NORTH' ) {
            $yc        = $top;
            $r         = $r_o;
            $delta     = $delta_o;
            $angle     = 270;
            $max_angle = 360;
        }
        else {
            $yc        = $bottom;
            $r         = $r_i;
            $delta     = -$delta_i;
            $angle     = 90;
            $max_angle = 0;
        }
    }
    elsif ( $d eq 'NORTH' && $y < $bottom ) {
        if ( $nd eq 'WEST' ) {
            $xc        = $left;
            $r         = $r_o;
            $delta     = $delta_o;
            $angle     = 0;
            $max_angle = 90;
        }
        else {
            $xc        = $right;
            $r         = $r_i;
            $delta     = -$delta_i;
            $angle     = 180;
            $max_angle = 90;
        }
        $yc = $bottom;
    }
    elsif ( $d eq 'SOUTH' && $y > $top ) {
        if ( $nd eq 'WEST' ) {
            $xc        = $left;
            $r         = $r_i;
            $delta     = -$delta_i;
            $angle     = 360;
            $max_angle = 270;
        }
        else {
            $xc        = $right;
            $r         = $r_o;
            $delta     = $delta_o;
            $angle     = 180;
            $max_angle = 270;
        }
        $yc = $top;
    }
    else {
        return;
    }

    my ( $fx, $fy, $fvx, $fvy );
    if ( $nd eq 'WEST' ) {
        $fx  = $left;
        $fy  = $top + 12;
        $fvx = -1;
        $fvy = 0;
    }
    elsif ( $nd eq 'EAST' ) {
        $fx  = $right;
        $fy  = $top + 38;
        $fvx = 1;
        $fvy = 0;
    }
    elsif ( $nd eq 'NORTH' ) {
        $fx  = $left + 38;
        $fy  = $top;
        $fvx = 0;
        $fvy = -1;
    }
    elsif ( $nd eq 'SOUTH' ) {
        $fx  = $left + 12;
        $fy  = $bottom;
        $fvx = 0;
        $fvy = 1;
    }

    $car->next_road( $self->next($d) );
    $car->turn(
        {   x         => $xc,
            y         => $yc,
            r         => $r,
            angle     => $angle,
            max_angle => $max_angle,
            delta     => $delta,
            finish    => {
                x         => $fx,
                y         => $fy,
                v_x       => $fvx,
                v_y       => $fvy,
                direction => $nd,
            },
        }
    );
}

sub contains {
    my ( $self, $x, $y ) = @_;

    my $map = $self->map;
    return
           $x >= $self->x * $map->tile_w
        && $x < ( $self->x + 1 ) * $map->tile_w
        && $y >= $self->y * $map->tile_h
        && $y < ( $self->y + 1 ) * $map->tile_h;
}

no Mouse;

1;


lib/Games/PuzzleCars/Sprite.pm
package Games::PuzzleCars::Sprite;
use Mouse;
use SDLx::Sprite::Animated;

has _sprite => (
    is       => 'ro',
    isa      => 'SDLx::Sprite::Animated',
    required => 1,
    handles  => [qw( x y h w rect clip draw draw_xy alpha_key )],
);

has increment => (
    is      => 'ro',
    isa     => 'Int',
    lazy    => 1,
    default => sub { 360 / ( $_[0]->rows * $_[0]->cols ) },
);

has rows => (
    is      => 'ro',
    isa     => 'Int',
    lazy    => 1,
    default => sub { $_[0]->h / $_[0]->rect->h },
);

has cols => (
    is      => 'ro',
    isa     => 'Int',
    lazy    => 1,
    default => sub { $_[0]->w / $_[0]->rect->w },
);

around BUILDARGS => sub {
    my ( $orig, $class, %args ) = @_;

    my $animated = SDLx::Sprite::Animated->new(
        rect    => $args{rect},
        surface => $args{surface},
    );
    $animated->alpha_key(0x1D1D1D);

    return $class->$orig( _sprite => $animated );
};

sub rotation {
    my ( $self, $rot ) = @_;

    while ( $rot >= 360 ) { $rot -= 360 }
    while ( $rot < 0 ) { $rot += 360 }

    my $frame = int( $rot / $self->increment );

    my $clip = $self->clip;
    $clip->x( ( $frame % $self->cols ) * $self->rect->h );
    $clip->y( int( $frame / $self->cols ) * $self->rect->w );
}

no Mouse;

1;


tools/create_rotatable_sprite.pl
#!/usr/bin/env perl
use strict;
use warnings;
use 5.010;
use Imager;

sub main {
    my ( $i_file, $o_file, $bg_color, $num ) = @_;

    die "Need an input file\n"  unless $i_file;
    die "Need an output file\n" unless $o_file;

    $bg_color //= '#1D1D1D';

    $num //= 360 / 5;
    my $deg = 360 / $num;

    my $input = Imager->new;
    $input->read( file => $i_file )
        or die "Could not read: $i_file: " . $input->errstr;

    my $size = size( $input->getwidth, $input->getheight );

    my $output
        = Imager->new( xsize => $size * $num, ysize => $size, channels => 4 );
    $output->box(
        xmin   => 0,
        ymin   => 0,
        xmax   => $size * $num,
        ymax   => $size,
        filled => 1,
        color  => $bg_color,
    );

    foreach my $i ( 0 .. $num - 1 ) {
        my $rot = $input->rotate( degrees => -$i * $deg, back => $bg_color );
        my $left = $i * $size + ( $size - $rot->getwidth ) / 2;
        my $top = ( $size - $rot->getheight ) / 2;
        $output->paste( src => $rot, left => $left, top => $top );
    }

    $output->write( file => $o_file )
        or die "Cannot save $o_file: ", $output->errstr;

    exit;
}

sub size {
    my ( $w, $h ) = @_;
    return int( sqrt( $w * $w + $h * $h ) ) + 1;
}

main(@ARGV);


lib/Games/PuzzleCars/Intersection.pm
package Games::PuzzleCars::Intersection;
use Mouse;

extends 'Games::PuzzleCars::Road';

has direction => (
    is      => 'rw',
    isa     => 'Int',
    default => 0,
);

has arrow => (
    is       => 'ro',
    isa      => 'SDLx::Sprite',
    required => 1,
);

sub _direction_count {
    my ($self) = @_;

    return scalar keys %{ $self->directions };
}

sub _next_direction {
    my ( $self, $direction ) = @_;

    return $direction if $self->direction == 2;

    my %next = (
        NORTH => [qw( EAST NORTH WEST SOUTH )],
        EAST  => [qw( SOUTH EAST NORTH WEST )],
        SOUTH => [qw( WEST SOUTH EAST NORTH )],
        WEST  => [qw( NORTH WEST SOUTH EAST )],
    );

    my @dirs
        = grep { defined $self->directions->{$_} } @{ $next{$direction} };

    if ( $self->_direction_count == 4 ) {
        @dirs = grep { $_ ne $direction } @dirs;
    }

    return $dirs[ $self->direction ];
}

sub handle_event {
    my ( $self, $event ) = @_;

    my $e_x = $event->button_x;
    my $e_y = $event->button_y;

    my $x      = ( $self->x + 0.5 ) * $self->map->tile_w;
    my $y      = ( $self->y + 0.5 ) * $self->map->tile_h;
    my $s      = $self->arrow->h;
    my $left   = $x - $s / 2;
    my $right  = $x + $s / 2;
    my $top    = $y - $s / 2;
    my $bottom = $y + $s / 2;

    return
        unless $e_x > $left && $e_x < $right && $e_y > $top && $e_y < $bottom;

    $self->direction(
        ( $self->direction + 1 ) % ( $self->_direction_count - 1 ) );
}

sub draw {
    my ( $self, $surface ) = @_;

    my $arrow = $self->arrow;
    my $x     = ( $self->x + 0.5 ) * $self->map->tile_w;
    my $y     = ( $self->y + 0.5 ) * $self->map->tile_h;
    my $s     = $arrow->h;
    $arrow->clip( [ $self->direction * $s, 0, $s, $s ] );
    $arrow->draw_xy( $surface, $x - $s / 2, $y - $s / 2 );
}

no Mouse;

1;


lib/Games/PuzzleCars.pm
package Games::PuzzleCars;
use strict;
use warnings;
use Mouse;
use namespace::clean -except => 'meta';
use FindBin;
use File::Spec;
use File::ShareDir;
use SDL 2.500;
use SDLx::App;
use SDLx::Surface;
use SDLx::Text;
use Games::PuzzleCars::Map;
use Games::PuzzleCars::Car;

has share_dir => (
    is      => 'ro',
    isa     => 'Str',
    builder => '_build_share_dir',
);

has app => (
    is      => 'ro',
    isa     => 'SDLx::App',
    lazy    => 1,
    builder => '_build_app',
    handles => [qw( run )],
);

has difficulty => (
    is       => 'ro',
    isa      => 'Str',
    required => 1,
);

has map => (
    is      => 'rw',
    isa     => 'Games::PuzzleCars::Map',
    lazy    => 1,
    builder => '_build_map',
);

has cars => (
    is      => 'ro',
    isa     => 'ArrayRef[Games::PuzzleCars::Car]',
    default => sub { [] },
);

has car_frequency => (
    is      => 'ro',
    isa     => 'Num',
    default => 3,
);

has _last_car_time => (
    is      => 'rw',
    isa     => 'Num',
    default => 0,
);

has _car_colors => (
    is      => 'ro',
    isa     => 'ArrayRef[Str]',
    default => sub { [qw( r g y )] },
);

has points => (
    is      => 'rw',
    isa     => 'Int',
    default => 0,
);

has _label => (
    is      => 'ro',
    isa     => 'SDLx::Text',
    default => sub {
        SDLx::Text->new( x => 10, y => 10, size => 32, color => [ 0, 0, 0 ] );
    },
);

has _surfaces => (
    is      => 'ro',
    isa     => 'HashRef[SDLx::Surface]',
    default => sub { [] },
    lazy    => 1,
    builder => '_build_surfaces',
);

sub _build_share_dir {
    my $root = File::Spec->catdir( $FindBin::Bin, '..' );
    if ( -f File::Spec->catfile( $root, 'dist.ini' ) ) {
        return File::Spec->catdir( $root, 'share' );
    }
    else {
        return File::ShareDir::dist_dir('Games-PuzzleCars');
    }
}

sub _build_app {
    return SDLx::App->new(
        w     => 800,
        h     => 600,
        dt    => 0.02,
        eoq   => 1,
        delay => 20,
        title => 'Puzzle Cars',
    );
}

sub _build_map {
    my $self = shift;
    return Games::PuzzleCars::Map->new(
        w    => $self->app->w,
        h    => $self->app->h,
        file => File::Spec->catfile(
            $self->share_dir, 'maps', $self->{difficulty} . '.txt'
        ),
        intersection => {
            w     => 40,
            h     => 40,
            image => File::Spec->catfile( $self->share_dir, 'arrows.bmp' ),

        },
        roads => {
            w       => 50,
            h       => 50,
            image   => File::Spec->catfile( $self->share_dir, 'roads.bmp' ),
            mapping => [
                [   [ [ 6, 0 ] ],
                    [ [ 2, 0 ], [ 2, 1 ] ],
                    [ [ 4, 0 ], [ 5, 0 ] ],
                    [ [ 0, 0 ] ],
                    [ [ 8, 0 ], [ 8, 1 ], [ 9, 0 ], [ 9, 1 ] ],
                ],
                [   [ [ 7, 0 ] ],
                    [ [ 3, 0 ], [ 3, 1 ] ],
                    [ [ 4, 0 ], [ 5, 0 ] ],
                    [ [ 1, 0 ] ],
                    [ [ 8, 0 ], [ 8, 1 ], [ 9, 0 ], [ 9, 1 ] ],
                ],
                [   [ [ 6, 1 ] ],
                    [ [ 2, 0 ], [ 2, 1 ] ],
                    [ [ 4, 1 ], [ 5, 1 ] ],
                    [ [ 0, 1 ] ],
                    [ [ 8, 0 ], [ 8, 1 ], [ 9, 0 ], [ 9, 1 ] ],
                ],
                [   [ [ 7, 1 ] ],
                    [ [ 3, 0 ], [ 3, 1 ] ],
                    [ [ 4, 1 ], [ 5, 1 ] ],
                    [ [ 1, 1 ] ],
                    [ [ 8, 0 ], [ 8, 1 ], [ 9, 0 ], [ 9, 1 ] ],
                ],
            ],
        },
    );
}

sub _build_surfaces {
    my ($self) = @_;

    my %surfaces;
    foreach my $color ( @{ $self->_car_colors } ) {
        $surfaces{ $color . '_car' }
            = SDLx::Surface->load(
            File::Spec->catfile( $self->share_dir, 'cars', $color . '.bmp' )
            );
    }

    return \%surfaces;
}

sub BUILD {
    my ($self) = @_;

    $self->_add_car();

    my $app = $self->app;
    $app->add_event_handler( sub { $self->handle_event(@_) } );
    $app->add_move_handler( sub  { $self->handle_move(@_) } );
    $app->add_show_handler( sub  { $self->handle_show(@_) } );
}

sub handle_event {
    my ( $self, $event, $app ) = @_;
    $self->map->handle_event($event);
}

sub handle_move {
    my ( $self, $step, $app, $t ) = @_;

    if ( $t > $self->_last_car_time + $self->car_frequency ) {
        $self->_last_car_time($t);
        $self->_add_car();
    }

    my $map  = $self->map;
    my @cars = @{ $self->cars };
    foreach my $car (@cars) {
        $car->move( $step, $app, $t );

        if (   $car->x < 0
            || $car->x > $map->w * $map->tile_w
            || $car->y < 0
            || $car->y > $map->h * $map->tile_h )
        {

            my $road = $car->road;
            if ( my $exit = $map->exits->[ $road->x ][ $road->y ] ) {
                if ( $car->color eq $exit ) {
                    $self->points( $self->points + 10 );
                }
                else {
                    $self->points( $self->points - 10 );
                }
            }

            @{ $self->cars } = grep { $_ ne $car } @{ $self->cars };
        }
    }
}

sub handle_show {
    my ( $self, $delta, $app ) = @_;

    $self->map->draw($app);
    $_->draw($app) foreach @{ $self->cars };
    $self->_label->write_to( $app, 'Score: ' . $self->points );
    $app->update();
}

sub _add_car {
    my ($self) = @_;

    my $left  = 37;
    my $right = 63;

    my $map     = $self->map;
    my $borders = $map->borders;
    my $border  = $borders->[ int rand @$borders ];

    my $colors = $self->_car_colors;
    my $color  = $colors->[ int rand @$colors ];

    my ( $x, $y, $rot, $v_x, $v_y, $direction );
    if ( $border->x == 0 && defined $border->directions->{WEST} ) {
        $x         = 0;
        $y         = $border->y * $map->tile_h + $right;
        $rot       = 0;
        $v_x       = 1;
        $v_y       = 0;
        $direction = 'EAST';
    }
    elsif ( $border->y == 0 && defined $border->directions->{NORTH} ) {
        $x         = $border->x * $map->tile_w + $left;
        $y         = 0;
        $rot       = 270;
        $v_x       = 0;
        $v_y       = 1;
        $direction = 'SOUTH';
    }
    elsif ( $border->x == $map->w - 1 && defined $border->directions->{EAST} )
    {
        $x         = $map->w * $map->tile_w;
        $y         = $border->y * $map->tile_h + $left;
        $rot       = 180;
        $v_x       = -1;
        $v_y       = 0;
        $direction = 'WEST';
    }
    elsif ( $border->y == $map->h - 1
        && defined $border->directions->{SOUTH} )
    {
        $x         = $border->x * $map->tile_w + $right;
        $y         = $map->h * $map->tile_h;
        $rot       = 90;
        $v_x       = 0;
        $v_y       = -1;
        $direction = 'NORTH';
    }

    die unless defined $x;

    my $car = Games::PuzzleCars::Car->new(
        color     => $color,
        rect      => SDL::Rect->new( 0, 0, 34, 34 ),
        surface   => $self->_surfaces->{ $color . '_car' },
        map       => $self->map,
        road      => $border,
        next_road => $border->next($direction),
        direction => $direction,
        x         => $x,
        y         => $y,
        rot       => $rot,
        v_x       => $v_x,
        v_y       => $v_y,
    );

    push @{ $self->cars }, $car;
}

__PACKAGE__->meta->make_immutable;

1;

=pod

=head1 SYNOPSIS

    use Games::PuzzleCars;

=head1 DESCRIPTION

What does this module do?

=head1 SEE ALSO

=over 4

=item * L<SDL>

=back

=cut

